### PR TITLE
Infer dataset change

### LINKED ISSUE #7859: Update Lesson and Exam model infer_dataset method to use the collection rather than assigned_by foreign key
When Lesson and Exam models were first created, their `infer_dataset` method referred to the user that it was assigned by using the `assigned_by` ForeignKey.

In a scenario where only a limited subset of a facility's data is synchronized to a device, such as in a single learner syncing scenario, this would cause Lessons and Quizzes to remain forever serialized in the Morango store, as the data for the assigning user would be absent, as outside the partition of data that would be synced for a single user.

In order to ensure that Lesson and Exam models can be properly deserialized, their `infer_dataset` method should instead rely on the `collection` ForeignKey, as the full Collection hierarchy of the Facility will be shared during a single user sync.

### PR DESCRIPTION
**Pull request by udithprabhu**
### Summary
Changed infet_dataset from Exam and Lesson from created_by to collection

…

### References
Fixes https://github.com/learningequality/kolibri/issues/7859

…

----

### Contributor Checklist

PR process:

- [X] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [X] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by jamalex** [comment]
Thanks for the contribution, @udithprabhu!

As you noted in the issue, a simple test case could be very helpful to add here. I would suggest perhaps adding it as a new testing class [in here](https://github.com/learningequality/kolibri/blob/release-v0.14.x/kolibri/core/auth/test/test_datasets.py). You'll need to create a Facility, a FacilityUser (linked to the Facility), a Classroom (linked to the Facility), and an Exam (linked to the Classroom and FacilityUser), and then you can call `cached_related_dataset_lookup` with both fields to compare them, as you say.

**Comment by udithprabhu** [comment]
> Thanks for the contribution, @udithprabhu!
> 
> As you noted in the issue, a simple test case could be very helpful to add here. I would suggest perhaps adding it as a new testing class [in here](https://github.com/learningequality/kolibri/blob/release-v0.14.x/kolibri/core/auth/test/test_datasets.py). You'll need to create a Facility, a FacilityUser (linked to the Facility), a Classroom (linked to the Facility), and an Exam (linked to the Classroom and FacilityUser), and then you can call `cached_related_dataset_lookup` with both fields to compare them, as you say.

I have made the changes as per your suggestions. I have listed all the conditions that I could think of. If I have missed any, please let me know.

**Review by jamalex — APPROVED** [review]
This looks great, thank you so much!

The failing test seems unrelated to these changes, so it shouldn't be an issue, but I'll wait for someone to confirm that before we do the final merge.

**Pull request merged**
